src/Outbox/MapBased.php

```
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Dsantang\DomainEventsDoctrine\Outbox;
+
+use Dsantang\DomainEvents\DomainEvent;
+use function get_class;
+
+final class MapBased extends EventsHandler
+{
+    /** @var converter[] */
+    private $conversionMap;
+
+    /**
+     * @param converter[] $conversionMap
+     */
+    public function __construct(OutboxMappedSuperclass $outboxMappedSuperclass, array $conversionMap)
```

One thing that we can improve in this constructor in the following.

Currently, if the map points to a class that does not implement the converter interface, we will blow up runtime.

But this can be solved at container compilation time, buy performing a check at the constructor level, in order to assert that every converter pointed by our `conversionMap` implements the `Converter` interface.

It it doesn't we can raise a specific exception that explain our `Client` about their misconfiguration.

---

good point! And we have to add this validation explicitly because the "splat" operator does not work with associative arrays (like `$conversionMap`)

---

Yup, and also cause we can explicitly say what the problem is to our Client by creating a custom exception!

---

How does this look like from a DIC perspective? how are we passing these instances into the array?

---

We would basically be passing this configuration via a parameter containing a collection of services, which keys are the class names.

---

Regarding this, maybe adding a builder method could also to add a converter can also help things out!

sumfin that enables us to do this:

```
<service id="namespace/MapBased" class="namespace/MapBased">
            <call method="addConverter">
                <argument>namespace/SpecificDomainEvent</argument>
                <argument type="service" id="namespace/SpecificDomainEventConverter"/>
            </call>
</service>
```

I like it cause it's very clean

---

Fuck, I actually think this is the only solution, cause Symfony does not allow that, so we cannot pass collections of services...

---

@fabiotc, we're only missing the builder method `addConverter(string $domainEventClass, Converter $converter)` and we're good :)

---

done!